File: mtl_examples.kt
// Kotlin MTL: Pragmatic safety with sealed classes
// GF(3) trit: 0 (ERGODIC)

sealed class MTLFormula {
    abstract fun evaluate(trace: List<TimedEvent>, currentTime: Long): Boolean
}

data class Interval(val lower: Long, val upper: Long)
data class TimedEvent(val timestamp: Long, val predicate: String, val value: Boolean)

data class Predicate(val name: String) : MTLFormula() {
    override fun evaluate(trace: List<TimedEvent>, currentTime: Long): Boolean =
        trace
            .filter { it.timestamp <= currentTime && it.predicate == name }
            .lastOrNull()
            ?.value ?: false
}

data class Eventually(val interval: Interval, val formula: MTLFormula) : MTLFormula() {
    override fun evaluate(trace: List<TimedEvent>, currentTime: Long): Boolean {
        val start = currentTime + interval.lower
        val end = currentTime + interval.upper
        return trace
            .filter { it.timestamp in start..end }
            .any { formula.evaluate(trace, it.timestamp) }
    }
}

data class Always(val interval: Interval, val formula: MTLFormula) : MTLFormula() {
    override fun evaluate(trace: List<TimedEvent>, currentTime: Long): Boolean {
        val start = currentTime + interval.lower
        val end = currentTime + interval.upper
        val eventsInInterval = trace.filter { it.timestamp in start..end }
        return eventsInInterval.isNotEmpty() && 
               eventsInInterval.all { formula.evaluate(trace, it.timestamp) }
    }
}

object CRDTProperties {
    fun convergence(tauMix: Long) = 
        Eventually(Interval(0, tauMix), Predicate("replicas_converged"))
    
    fun commutativity() = 
        Always(Interval(0, Long.MAX_VALUE), Predicate("merge_commutative"))
}

File: mtl_examples.scala
// Scala MTL: Functional abstractions with case classes
// GF(3) trit: +1 (PLUS generator)

sealed trait MTLFormula {
  def evaluate(trace: Seq[TimedEvent], currentTime: Long): Boolean
}

case class Interval(lower: Long, upper: Long)
case class TimedEvent(timestamp: Long, predicate: String, value: Boolean)

case class Predicate(name: String) extends MTLFormula {
  def evaluate(trace: Seq[TimedEvent], currentTime: Long): Boolean =
    trace
      .filter(_.timestamp <= currentTime)
      .filter(_.predicate == name)
      .lastOption
      .map(_.value)
      .getOrElse(false)
}

case class Eventually(interval: Interval, formula: MTLFormula) extends MTLFormula {
  def evaluate(trace: Seq[TimedEvent], currentTime: Long): Boolean = {
    val start = currentTime + interval.lower
    val end = currentTime + interval.upper
    trace
      .filter(e => e.timestamp >= start && e.timestamp <= end)
      .exists(e => formula.evaluate(trace, e.timestamp))
  }
}

case class Always(interval: Interval, formula: MTLFormula) extends MTLFormula {
  def evaluate(trace: Seq[TimedEvent], currentTime: Long): Boolean = {
    val start = currentTime + interval.lower
    val end = currentTime + interval.upper
    val eventsInInterval = trace.filter(e => e.timestamp >= start && e.timestamp <= end)
    eventsInInterval.nonEmpty && eventsInInterval.forall(e => formula.evaluate(trace, e.timestamp))
  }
}

object CRDTProperties {
  def convergence(tauMix: Long) = 
    Eventually(Interval(0, tauMix), Predicate("replicas_converged"))
  
  def commutativity = 
    Always(Interval(0, Long.MaxValue), Predicate("merge_commutative"))
  
  def associativity = 
    Always(Interval(0, Long.MaxValue), Predicate("merge_associative"))
}
